ln-gateway/src/cln.rs

```diff
@@ -40,3 +57,197 @@ pub struct HtlcAccepted {
+}
+
+async fn withdraw_rpc(
+    plugin: Plugin<GatewayRpcSender>,
+    value: serde_json::Value,
+) -> Result<serde_json::Value, Error> {
+    let withdraw: WithdrawPayload = serde_json::from_value(value)?;
+    let txid = plugin.state().send(withdraw).await?;
+    Ok(json!({ "fedimint_txid": txid.to_string() }))
+}
+
+pub async fn build_cln_rpc(sender: GatewayRpcSender) -> Result<LnRpcRef, Error> {
+    if let Some(plugin) = Builder::new(sender, stdin(), stdout())
+        .option(options::ConfigOption::new(
+            "fedimint-cfg",
+            // FIXME: cln_plugin doesn't support parameters without defaults
+            options::Value::String("default-dont-use".into()),
+            "fedimint config directory",
+        ))
+        .option(options::ConfigOption::new(
+            "fedimint-host",
+            options::Value::String("127.0.0.1".into()),
+            "gateway hostname",
+        ))
+        .option(options::ConfigOption::new(
+            "fedimint-port",
+            options::Value::String("8080".into()),
+            "gateway port",
+        ))
+        .rpcmethod("gw-balance", "Display ecash token balance", balance_rpc)
+        .rpcmethod(
+            "gw-deposit",
+            "Deposit into federation. Args: <txoutproof> <bitcoin-transaction>",
+            deposit_rpc,
+        )
+        .rpcmethod(
+            "gw-withdraw",
+            "Withdraw from federation. Args: <address> <sats>",
+            withdraw_rpc,
+        )
+        .rpcmethod("gw-address", "Generate deposit address", address)
+        .hook("htlc_accepted", |plugin, value| async move {
+            // This callback needs to be `Sync`, so we use tokio::spawn
+            let handle = tokio::spawn(async move {
+                htlc_accepted_hook(plugin, value).await.or_else(|e| {
+                    error!("htlc_accepted error {:?}", e);
+                    // cln_plugin doesn't handle errors very well ... tell it to proceed normally
+                    Ok(json!({ "result": "continue" }))
+                })
+            });
+            handle.await?
+        })
+        .dynamic() // Allow reloading the plugin
+        .start()
+        .await?
+    {
+        let work_dir = match plugin.option("fedimint-cfg") {
+            Some(options::Value::String(workdir)) => {
+                // FIXME: cln_plugin doesn't yet support optional parameters
+                if &workdir == "default-dont-use" {
+                    panic!("fedimint-cfg option missing")
+                } else {
+                    PathBuf::from(workdir)
+                }
+            }
+            _ => unreachable!(),
+        };
+
+        let host = match plugin.option("fedimint-host") {
+            Some(options::Value::String(host)) => host,
+            _ => unreachable!(),
+        };
+        let port = match plugin.option("fedimint-port") {
+            Some(options::Value::String(port)) => port,
+            _ => unreachable!(),
```

Comment:
In further refactorings these should probably become errors or more descriptive panics.

Reply:
💯